core/src/main/java/io/grpc/transport/MessageDeframer.java

```diff
@@ -214,67 +322,82 @@ private void checkNotClosed() {
+        try {
+          return current.stream;
+        } finally {
+          current = null;
         }
       }
-      // We are stalled when there are no more bytes to process. This allows delivering errors as
-      // soon as the buffered input has been consumed, independent of whether the application
-      // has requested another message.
-      stalled = !isDataAvailable();
-
-      if (endOfStream) {
-        if (!isDataAvailable()) {
-          listener.endOfStream();
-        } else if (stalled) {
-          // We've received the entire stream and have data available but we don't have
-          // enough to read the next frame ... this is bad.
-          throw Status.INTERNAL.withDescription("Encountered end-of-stream mid-frame")
-              .asRuntimeException();
+
+      @Override
+      public boolean hasNext() {
+        int consumedBytes = 0;
+        Frame peek;
+        try {
+          if (current != null) {
+            return true;
+          }
+          do {
+            peek = frameQueue.peek();
+            if (peek == null) {
+              return false;
+            }
+            if (peek == END_OF_STREAM) {
+              // Transport enqueued EOS, so empty queue and report to the listener
+              frameQueue.remove();
+              listener.endOfStream();
+              return false;
+            }
+            if (applicationCount <= 0) {
+              if (peek.stream != null) {
+                return false;
+              }
+            }
+            current = frameQueue.remove();
+            consumedBytes += current.consumedBytes;
+            if (current.stream == null) {
+              current = null;
+            } else {
+              applicationCount--;
+              return true;
+            }
+          } while (true);
+        } finally {
+          if (consumedBytes > 0) {
+            listener.bytesRead(consumedBytes);
+          }
         }
       }
 
-      // Never indicate that we're stalled if we've received all the data for the stream.
-      stalled &= !endOfStream;
+      @Override
+      public void remove() {
+        throw new UnsupportedOperationException();
+      }
 
-      // If we're transitioning to the stalled state, notify the listener.
-      boolean previouslyStalled = deliveryStalled;
-      deliveryStalled = stalled;
-      if (stalled && !previouslyStalled) {
-        listener.deliveryStalled();
+      @Override
+      public void forEachRemaining(Consumer action) {
+        throw new UnsupportedOperationException();
       }
-    } finally {
-      inDelivery = false;
-    }
+    };
+    // Notify the application, beyond this point any consumption of the iterator is considered
+    // to be done on an application thread
+    listener.messagesAvailable(messages);
   }
 
   private boolean isDataAvailable() {
-    return unprocessed.readableBytes() > 0 || (nextFrame != null && nextFrame.readableBytes() > 0);
+    return unprocessed.readableBytes() > 0;
```

Comment:
This is only used by `enqueue` in a couple of places ... maybe get rid of the method and just do this inline? Might make the code more clear.